cmd/util/ledger/migrations/account_based_migration.go

```diff
@@ -0,0 +1,210 @@
+		return "", false, nil
+	}
+	return id.Owner, true, nil
+}
+
+// PayloadGroup groups payloads by account.
+// For global payloads, it's stored under NonAccountPayloads field
+type PayloadGroup struct {
+	NonAccountPayloads []ledger.Payload
+	Accounts           map[string][]ledger.Payload
+}
+
+// PayloadGrouping is a reducer function that adds the given payload to the corresponding
+// group under its account
+func PayloadGrouping(groups *PayloadGroup, payload ledger.Payload) (*PayloadGroup, error) {
+	address, isAccount, err := PayloadToAccount(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if isAccount {
+		groups.Accounts[address] = append(groups.Accounts[address], payload)
+	} else {
+		groups.NonAccountPayloads = append(groups.NonAccountPayloads, payload)
+	}
+
+	return groups, nil
+}
+
+// AccountMigrator takes all the payloads that belong to the given account
+// and return the migrated payloads
+type AccountMigrator func(account string, payloads []ledger.Payload, pathFinderVersion uint8) ([]ledger.Payload, error)
+
+// MigrateByAccount teaks a migrator function and all the payloads, and return the migrated payloads
+func MigrateByAccount(migrator AccountMigrator, allPayloads []ledger.Payload, pathFinderVersion uint8) (
+	[]ledger.Payload, []ledger.Path, error) {
+	groups := &PayloadGroup{
+		NonAccountPayloads: make([]ledger.Payload, 0),
+		Accounts:           make(map[string][]ledger.Payload),
+	}
+
+	log.Info().Msgf("start grouping for a total of %v payloads", len(allPayloads))
+
+	var err error
+	logGrouping := util.LogProgress("grouping payload", len(allPayloads), &log.Logger)
+	for i, payload := range allPayloads {
+		groups, err = PayloadGrouping(groups, payload)
+		if err != nil {
+			return nil, nil, err
+		}
+		logGrouping(i)
+	}
+
+	log.Info().Msgf("finish grouping for payloads by account: %v groups in total, %v NonAccountPayloads",
+		len(groups.Accounts), len(groups.NonAccountPayloads))
+
+	// migrate the payloads under accounts
+	// migrated, paths, err := MigrateGroupSequentially(migrator, groups.Accounts, pathFinderVersion)
+	migrated, paths, err := MigrateGroupConcurrently(migrator, groups.Accounts, pathFinderVersion)
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not migrate group: %w", err)
+	}
+
+	log.Info().Msgf("finished migrating payloads for %v account", len(groups.Accounts))
+
+	nonAccountPaths, err := pathfinder.PathsFromPayloads(groups.NonAccountPayloads, pathFinderVersion)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not find paths for non account payloads: %w", err)
+	}
+
+	// add the non accounts which don't need to be migrated
+	migrated = append(migrated, groups.NonAccountPayloads...)
+	paths = append(paths, nonAccountPaths...)
+
+	log.Info().Msgf("finished migrating all account based payloads, total migrated payloads: %v", len(migrated))
+
+	return migrated, paths, nil
+}
+
+// MigrateGroupSequentially migrate the payloads in the given payloadsByAccount map which
```

Comment:
dead code ...

Reply:
Will be using this code to benchmark against the concurrent version when real migration is ready

Reply:
in that case, at least rm the commented out line of code at line 87.

if you refactor this into an interface, say GroupMigrator, you can switch out MigrateGroupConcurrently for MigrateGroupSequentially